sources/csharp2/KalturaClient/Request/BaseRequestBuilder.cs

```diff
@@ -75,231 +65,231 @@ public BaseRequestBuilder<T> SetCompletion(OnErrorHandler errorHandler)
             return this;
         }
 
+        public virtual Params getParameters(bool includeServiceAndAction)
+        {
+            Params kparams = new Params();
+
+            if (client != null)
+                kparams.Add(client.RequestConfiguration.ToParams(false));
+
+            kparams.Add(base.ToParams(false));
+
+            if (includeServiceAndAction)
+                kparams.Add("service", service);
+
+            return kparams;
+        }
+
+        public virtual Files getFiles()
+        {
+            return new Files();
+        }
+
         public virtual BaseRequestBuilder<T> Build(Client client)
         {
             this.client = client;
             Boundary = "---------------------------" + DateTime.Now.Ticks.ToString("x");
             return this;
         }
-
-        public void Execute(Client client = null)
+        
+        public virtual void OnComplete(object response, Exception error)
         {
-            startTime = DateTime.Now;
-
-            if (this.client == null)
+            if (onCompletion != null)
             {
-                this.Build(client);
+                onCompletion((T)response, error);
             }
-
-            if (this.client == null)
+            if (onError != null && error != null)
             {
-                OnComplete(null, new ClientException("No client instance defined"));
+                onError(error);
             }
+        }
 
-            string url = client.Configuration.ServiceUrl + "/api_v3" + getPath();
-            this.Log("url: [" + url + "]");
+        public async Task<T> ExecuteAsync(Client client = null)
+        {
+            var sw = new Stopwatch();
+            sw.Start();
 
-            // build request
-            request = (HttpWebRequest)HttpWebRequest.Create(url);
-            if (getFiles().Count == 0)
-            {
-                request.Timeout = client.Configuration.Timeout;
-            }
-            else
-            {
-                request.Timeout = Timeout.Infinite;
-            }
-            request.Method = "POST";
+            if (this.client == null) { this.Build(client); }
+            if (this.client == null) { OnComplete(null, new ClientException("No client instance defined")); }
 
-            request.Headers = getHeaders();
-            request.AutomaticDecompression = DecompressionMethods.GZip | DecompressionMethods.Deflate;
-            request.Accept = "application/xml";
-            request.ContentType = getContentType();
+            var url = client.Configuration.ServiceUrl + "/api_v3" + getPath();
+
+            var requestId = (Interlocked.Increment(ref Client.REQUEST_COUNTER)).ToString("X5");
```

Comment:
requestId could be initiated in the constructor.

Reply:
done: 4025cc3858f2210933ab5c6759e40afab0680c4c